### PR TITLE
Enhancement to binned_statistic: option to unraveled returned bin-number mappings, and able to pass multiple data arrays at once

### PR DESCRIPTION
**Pull request by lzkelley**
This PR addresses two issues:  
-   [Scipy 5449](https://github.com/scipy/scipy/issues/5449): `binned_statistic_2d` and `binned_statistic_dd` return mappings from the input data arrays to the bins they are placed in.  The mapping used (row-major) linearized bin indices, without documentation.  The enhancements in this PR both improve the documentation for the same behavior, and provide an input argument (`expand_binnumbers = False`) which, when True, unravels the bin-number mapping into an ndarray where each row gives the bins for each corresponding dimension.
-   [Numpy 4718](https://github.com/numpy/numpy/issues/4718): It was suggested that it would be useful for histogram to operate on numerous 'weightings' simulataneously (e.g. when calculating histogrammed moments).  For example, Previously:
  `stats_v, edges, bins = scipy.stats.binned_statisic(x, v, statistic='mean')`
  `stats_w, edges, bins = scipy.stats.binned_statisic(x, w, statistic='mean')`
  New:
  `(stats_v, stats_w), edges, bins = scipy.stats.binned_statisic(x, (v, w), statistic='mean')`

The documentation has been expanded to reflect the new functionality, and tests have been added.  Additionally, I've changed some of the variable names for consistency (e.g. using `x_edges` instead of `x_edge` to be consistent with `bin_edges`, etc), and for clarity (e.g. changed `sampBin` instead of `Ncount` for a variable which stores which bin each sample belongs in).

### CODE REVIEW

**Comment by jakevdp** [comment]
Hi – thanks for taking this on! The red X's above mean there are some test failures. You can click on it to see the specific tests that failed. From a cursory glance, it looks like there's a function in there with a forgotten ``return`` statement.

File: scipy/stats/_binned_statistic.py [hunk 1]
```diff
@@ -12,26 +12,28 @@
 
 
 BinnedStatisticResult = namedtuple('BinnedStatisticResult',
-                                   ('statistic', 'bin_edges', 'binnumber'))
+                                   ('statistic', 'bin_edges', 'binnumbers'))
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: This is a backward-incompatible change; let's leave the names the same as before so we don't need a deprecation cycle.

File: scipy/stats/_binned_statistic.py [hunk 2]
```diff
@@ -12,26 +12,28 @@
...
 
     Parameters
     ----------
-    x : array_like
+    x : (N,) array_like
         A sequence of values to be binned.
-    values : array_like
-        The values on which the statistic will be computed.  This must be
-        the same shape as `x`.
+    values : (N,) array_like or set of (N,) array_like
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: Not actually a *set* of arrays, but a *list* of arrays, yes?

File: scipy/stats/_binned_statistic.py [hunk 3]
```diff
@@ -56,7 +58,8 @@ def binned_statistic(x, values, statistic='mean',
         bin edges, including the rightmost edge, allowing for non-uniform bin
         widths.  Values in `x` that are smaller than lowest bin edge are
         assigned to bin number 0, values beyond the highest bin are assigned to
-        ``bins[-1]``.
+        ``bins[-1]``.  If the bin edges are specified, the number of bins will
+        be, (nx = len(bins)-1).
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: indentation

File: scipy/stats/_binned_statistic.py [hunk 4]
```diff
@@ -90,7 +94,22 @@ def binned_statistic(x, values, statistic='mean',
...
+    First some basic examples:
+
+    Create two evenly spaced bins in the range of the given sample, and sum the
+    corresponding values in each of those bins:
+
+    >>> values = [1.0, 1.0, 2.0, 1.5, 3.0]
+    >>> stats.binned_statistic([1, 1, 2, 5, 7], values, 'sum', bins=2)
+    (array([ 4. ,  4.5]), array([ 1.,  4.,  7.]), array([1, 1, 1, 2, 2]))
+
+    Multiple sets of values can also be passed.  The statistic is calculated on
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: Multiple sequences or multiple arrays

File: scipy/stats/_binned_statistic.py [hunk 5]
```diff
@@ -152,36 +171,38 @@ def binned_statistic(x, values, statistic='mean',
...
 
-    return BinnedStatisticResult(medians, edges[0], xy)
+    return BinnedStatisticResult(medians, edges[0], binnumbers)
 
 
 BinnedStatistic2dResult = namedtuple('BinnedStatistic2dResult',
-                                     ('statistic', 'x_edge', 'y_edge',
-                                      'binnumber'))
+                                     ('statistic', 'x_edges', 'y_edges',
+                                      'binnumbers'))
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: Again, changing names is not backward compatible. Let's keep the old names.

File: scipy/stats/_binned_statistic.py [hunk 6]
```diff
@@ -203,38 +224,93 @@ def binned_statistic_2d(x, y, values, statistic='mean',
...
+        If 'False' (default): The returned `binnumbers` is a shape (N,) array
+        of linearized indices mapping each element of `sample` to its
+        corresponding bin (using row-major ordering).
+        If 'True': The returned `binnumbers` is a shape (2,N) ndarray where
+        each row indicates where the elements of `sample` should be inserted,
+        into the `x_edges` and `y_edges` arrays respectively, so as to keep
+        the arrays sorted.  In each dimension, a binnumber of `i` means the
+        corresponding value is between (D_edges[i-1], D_edges[i]), where 'D' is
+        either 'x' or 'y'.
+
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: This strikes me as a bit long for a parameter description. I would put these details under ``Notes`` below.

File: scipy/stats/_binned_statistic.py [hunk 7]
```diff
@@ -295,31 +374,50 @@ def binned_statistic_dd(sample, values, statistic='mean',
...
+        `expand_binnumbers` argument.
+        If 'False' (default): The returned `binnumbers` is a shape (N,) array
+        of linearized indices mapping each element of `sample` to its
+        corresponding bin (using row-major ordering).
+        If 'True': The returned `binnumbers` is a shape (D,N) ndarray where
+        each row indicates where the elements of `sample` should be inserted
+        into the corresponding `bin_edges` array so as to keep the array
+        sorted.  In each dimension, a binnumber of `i` means the
+        corresponding value is between (D_edges[i-1], D_edges[i]), where 'D' is
+        either 'x' or 'y'.
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: Again probably move this to notes.

File: scipy/stats/_binned_statistic.py [hunk 8]
```diff
@@ -331,36 +429,51 @@ def binned_statistic_dd(sample, values, statistic='mean',
...
-    nbin = np.empty(D, int)
-    edges = D * [None]
-    dedges = D * [None]
+    try:
+        # `values` is an ND-array.
+        Vdim, Vlen = values.shape
+    except (AttributeError, ValueError):
+        # Sample is a sequence of 1D arrays.
+        values = np.atleast_2d(values)
+        Vdim, Vlen = values.shape
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: You could just call ``np.atleast_2d`` from the beginning and skip the try/except for simplicity. If the array is already 2D, it's a no-op.

File: scipy/stats/_binned_statistic.py [hunk 9]
```diff
@@ -446,21 +564,31 @@ def binned_statistic_dd(sample, values, statistic='mean',
...
+    # Unravel binnumbers into an ndarray, each row the bins for each dimension
+    if(expand_binnumbers and Ndim > 1):
+        binnumbers = np.asarray(np.unravel_index(binnumbers, nbin))
+
+    if np.any(result.shape[1:] != nbin - 2):
         raise RuntimeError('Internal Shape Error')
 
-    return BinnedStatisticddResult(result, edges, xy)
+    # Remove excess dimension-zero if ``Vdim == 1``
+    result = result.squeeze()
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: This will produce an unexpected result if the user inputs a length-1 list of arrays. I think we need to save the input shape above, and use that here to choose the right output.

File: scipy/stats/tests/test_binned_statistic.py [hunk 10]
```diff
@@ -105,6 +106,20 @@ def test_1d_range_keyword(self):
...
+        w = self.w
+
+        stat1v, edges1v, bc1v = binned_statistic(x, v, 'mean', bins=10)
+        stat1w, edges1w, bc1w = binned_statistic(x, w, 'mean', bins=10)
+        stat2, edges2, bc2 = binned_statistic(x, [v, w], 'mean', bins=10)
+
+        assert_array_almost_equal(stat2[0], stat1v)
+        assert_array_almost_equal(stat2[1], stat1w)
+        assert_array_almost_equal(edges1v, edges2)
+        assert_array_almost_equal(bc1v, bc2)
```
Comment (jakevdp) [inline scipy/stats/tests/test_binned_statistic.py]: I think ``assert_allclose`` is preferred these days, but I'm not certain...

**Comment by jakevdp** [comment]
Overall this looks really good – nice work! Hopefully my comments above are clear.

File: scipy/stats/_binned_statistic.py (hunk 9 above)
Comment (lzkelley) [inline scipy/stats/_binned_statistic.py]: I'm not sure I understand what you mean by 'length-1'.  Because I'm doing ``values = np.atleast_2d(values)``, if the user inputs a `(N,)` array, they would get back a `(1,M)` array (for `N` length input arrays, and `M` bins) --- so I was squeezing that out.  Are you saying that if they input a `(1,N)` array *on purpose*, then they will erroneously get back a `(M,)` array?

To resolve my current understanding of the issue, I've switch to storing the initial number of dimensons, ``initVdim = np.ndim(values)`` in the beginning, and then at the end doing,  

    if(initVdim == 1):
        result = result.reshape(*(nbin-2))

Where `nbin` is actually the number of bins in each dimension `+2` for outlier bins.

File: scipy/stats/_binned_statistic.py (hunk 3 above)
Comment (lzkelley) [inline scipy/stats/_binned_statistic.py]: Sorry, don't understand what you mean here.

File: scipy/stats/_binned_statistic.py (hunk 2 above)
Comment (lzkelley) [inline scipy/stats/_binned_statistic.py]: Right, I didn't think about how `set` is a particular type.  At the same time, this could be an `ndarray` instead of a "list" *per se*.  Is "list" still the best terminology?

File: scipy/stats/_binned_statistic.py (hunk 1 above)
Comment (lzkelley) [inline scipy/stats/_binned_statistic.py]: Sure, that makes sense.  I've changed these back.  I guess the general principle is that a change should be above some threshold of improvement to warrant a deprecation?

File: scipy/stats/tests/test_binned_statistic.py (hunk 10 above)
Comment (lzkelley) [inline scipy/stats/tests/test_binned_statistic.py]: Yeah, [looks like](http://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_array_almost_equal.html) you're right!

File: scipy/stats/_binned_statistic.py (hunk 9 above)
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: >  Are you saying that if they input a (1,N) array on purpose, then they will erroneously get back a (M,) array?

Exactly. I would fix it by doing
```
values = np.asarray(values)
original_shape = values.shape
values = np.atleast_2d(values)
# ...
result = result.reshape(original_shape[:-1] + [M])
```
or something similar to that.

File: scipy/stats/_binned_statistic.py (hunk 3 above)
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: I guess it's OK – sorry!

**Comment by lzkelley** [comment]
All of the above issues have been addressed, and all checks are passing.  I think we're good to go!

File: scipy/stats/_binned_statistic.py [hunk 11]
```diff
@@ -90,11 +94,27 @@ def binned_statistic(x, values, statistic='mean',
...
+    >>> values = [[1.0, 1.0, 2.0, 1.5, 3.0], [2.0, 2.0, 4.0, 3.0, 6.0]]
+    >>> stats.binned_statistic([1, 1, 2, 5, 7], values, 'sum', bins=2)
+    (array([[ 4. ,  4.5], [ 8. ,  9. ]]), array([ 1.,  4.,  7.]),
+        array([1, 1, 1, 2, 2]))
 
     >>> stats.binned_statistic([1, 2, 1, 2, 4], np.arange(5), statistic='mean',
     ...                        bins=3)
-    (array([ 1.,  2.,  4.]), array([ 1.,  2.,  3.,  4.]), array([1, 2, 1, 2, 3]))
+    (array([ 1.,  2.,  4.]), array([ 1.,  2.,  3.,  4.]),
+        array([1, 2, 1, 2, 3]))
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: Does this pass doctests? Changing indentation can lead to issues.

Comment (lzkelley) [inline scipy/stats/_binned_statistic.py]: Yeah, the doctests passed as-well.

File: scipy/stats/_binned_statistic.py [hunk 12]
```diff
@@ -331,46 +447,60 @@ def binned_statistic_dd(sample, values, statistic='mean',
...
     except (AttributeError, ValueError):
-        # Sample is a sequence of 1D arrays.
+        # `sample` is a sequence of 1D arrays.
         sample = np.atleast_2d(sample).T
-        N, D = sample.shape
+        Dlen, Ndim = sample.shape
+
+    # Store initial shape of `values` to preserve it in the output
+    values = np.asarray(values)
+    inValuesShape = list(values.shape)
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: minor stylistic point, but we tend to avoid CamelCase for variable names. I'd use ``input_shape`` or something like that.

File: scipy/stats/_binned_statistic.py [hunk 13]
```diff
@@ -381,60 +511,65 @@ def binned_statistic_dd(sample, values, statistic='mean',
...
     if statistic == 'mean':
         result.fill(np.nan)
-        flatcount = np.bincount(xy, None)
-        flatsum = np.bincount(xy, values)
+        flatcount = np.bincount(binnumbers, None)
         a = flatcount.nonzero()
-        result[a] = flatsum[a] / flatcount[a]
+        for vv in xrange(Vdim):
+            flatsum = np.bincount(binnumbers, values[vv])
+            result[vv, a] = flatsum[a] / flatcount[a]
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: Why did you change the broadcasting approach to a ``for`` loop here? I'd assume the old version would be faster.

File: scipy/stats/_binned_statistic.py [hunk 14]
```diff
@@ -381,60 +511,65 @@ def binned_statistic_dd(sample, values, statistic='mean',
...
-        flatsum2 = np.bincount(xy, values ** 2)
+        flatcount = np.bincount(binnumbers, None)
         a = flatcount.nonzero()
-        result[a] = np.sqrt(flatsum2[a] / flatcount[a]
-                            - (flatsum[a] / flatcount[a]) ** 2)
+        for vv in xrange(Vdim):
+            flatsum = np.bincount(binnumbers, values[vv])
+            flatsum2 = np.bincount(binnumbers, values[vv] ** 2)
+            result[vv, a] = np.sqrt(flatsum2[a] / flatcount[a] -
+                                    (flatsum[a] / flatcount[a]) ** 2)
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: Same here – why the for loop?

File: scipy/stats/_binned_statistic.py [hunk 15]
```diff
@@ -381,60 +511,65 @@ def binned_statistic_dd(sample, values, statistic='mean',
...
+        result[:, a] = flatcount[np.newaxis, :]
     elif statistic == 'sum':
         result.fill(0)
-        flatsum = np.bincount(xy, values)
-        a = np.arange(len(flatsum))
-        result[a] = flatsum
+        for vv in xrange(Vdim):
+            flatsum = np.bincount(binnumbers, values[vv])
+            a = np.arange(len(flatsum))
+            result[vv, a] = flatsum
```
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: Same

File: scipy/stats/_binned_statistic.py (hunk 13 above)
Comment (jakevdp) [inline scipy/stats/_binned_statistic.py]: Oh, nevermind. It's for the multiple input arrays. Got it.

**Comment by jakevdp** [comment]
Looks great! I'm +1 for merge once these tests pass.

**Comment by rgommers** [comment]
Merged with some cleanup of the history in 6dbde4b. Thanks @lzkelley, @jakevdp

**Comment by nvaytet** [comment]
Was the ability to operate on numerous 'weightings' simulataneously only added to `binned_statistic` and not `binned_statistic_2d` and `binned_statistic_dd`?
